packages/common/solana/utils.ts

```diff
@@ -125,72 +142,125 @@ export async function signTransaction(invoker: Keypair | SignerWalletAdapter, tx
+    throw Error("Error with transaction parameters.");
+  }
+
+  for (let i = 0; i < 3; i++) {
+    let res: RpcResponseAndContext<SimulatedTransactionResponse>;
+    if (isTransactionVersioned(tx)) {
+      res = await connection.simulateTransaction(tx);
+    } else {
+      res = await connection.simulateTransaction(tx);
+    }
+    if (res.value.err) {
+      const errMessage = JSON.stringify(res.value.err);
+      if (!errMessage.includes("BlockhashNotFound") || i === 2) {
+        throw new SendTransactionError("failed to simulate transaction: " + errMessage, res.value.logs || undefined);
+      }
+    }
+    break;
+  }
 
-  if (!hash.lastValidBlockHeight || !tx.signature || !hash.blockhash) throw Error("Error with transaction parameters.");
+  const isVersioned = isTransactionVersioned(tx);
 
-  const signature = bs58.encode(tx.signature);
-  const confirmationStrategy: BlockheightBasedTransactionConfirmationStrategy = {
-    lastValidBlockHeight: hash.lastValidBlockHeight + 50,
-    signature,
-    blockhash: hash.blockhash,
-  };
-  try {
-    return await sendAndConfirmRawTransaction(connection, rawTx, confirmationStrategy);
-  } catch (e) {
-    // If BlockHeight expired, we will check tx status one last time to make sure
-    if (e instanceof TransactionExpiredBlockheightExceededError) {
-      await sleep(1000);
+  let signature: string;
+  if (isVersioned) {
+    signature = bs58.encode(tx.signatures[0]);
+  } else {
+    signature = bs58.encode(tx.signature!);
+  }
+
+  let blockheight = await connection.getBlockHeight(commitment);
+  let transactionSent = false;
+  const rawTransaction = tx.serialize();
+  while (blockheight < hash.lastValidBlockHeight) {
+    try {
+      await connection.sendRawTransaction(rawTransaction, {
+        maxRetries: 0,
+        minContextSlot: context.slot,
+        preflightCommitment: commitment,
+        skipPreflight: true,
+      });
+      transactionSent = true;
+    } catch (e) {
+      if (
+        transactionSent ||
+        (e instanceof SendTransactionError && e.message.includes("Minimum context slot has not been reached"))
+      ) {
+        continue;
+      }
+      throw e;
+    }
+    await sleep(500);
+    try {
       const value = await confirmAndEnsureTransaction(connection, signature);
-      if (!value) {
+      if (value) {
+        return signature;
+      }
+    } catch (e) {
+      if (e instanceof TransactionFailedError) {
         throw e;
       }
-      return signature;
+      await sleep(500);
+    }
+
+    try {
+      blockheight = await connection.getBlockHeight(commitment);
+    } catch (_e) {
+      await sleep(500);
     }
-    throw e;
   }
+
+  throw new Error(`Transaction ${signature} expired.`);
 }
 
 /**
  * Confirms and validates transaction success once
  * @param connection - Solana client connection
  * @param signature - Transaction signature
+ * @param passError - return status even if tx failed
```

Comment:
I find naming a bit strange, what should passError mean? Could be just `successful`?

Reply:
Or maybe better, `ignoreError`, if we are just printing it ant not throwing